File: frontend/src/api/reportExport.ts
export type ReportExportType =
  | "dashboard"
  | "funnel"
  | "source_performance"
  | "content_effectiveness";

export type ReportExportFormat = "csv" | "printable";

export type ReportExportQuery = {
  report_type: ReportExportType;
  format: ReportExportFormat;
  preset?: "last_7_days" | "last_30_days" | "this_month";
  start?: string;
  end?: string;
  grouping?: string;
};

export async function downloadReportExport(params: ReportExportQuery): Promise<void> {
  const sp = new URLSearchParams();
  sp.set("report_type", params.report_type);
  sp.set("format", params.format);
  if (params.preset) sp.set("preset", params.preset);
  if (params.start) sp.set("start", params.start);
  if (params.end) sp.set("end", params.end);
  if (params.grouping) sp.set("grouping", params.grouping);

  const r = await fetch(`/reports/export?${sp.toString()}`);
  if (!r.ok) {
    const body = await r.json().catch(() => ({}));
    throw new Error((body as { detail?: string }).detail ?? "report export failed");
  }
  const blob = await r.blob();
  const disposition = r.headers.get("Content-Disposition") ?? "";
  const match = /filename="([^"]+)"/.exec(disposition);
  const filename = match?.[1] ?? `report-export.${params.format === "csv" ? "csv" : "html"}`;
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  a.click();
  URL.revokeObjectURL(url);
}

File: src/livelead/interfaces/rest/report_export.py
"""Report export download endpoint (US-019)."""

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from livelead.application.reporting.export_service import ReportExportService
from livelead.domain.reporting.content_effectiveness import InvalidContentGrouping
from livelead.domain.reporting.report_export import UnsupportedReportExport
from livelead.domain.reporting.source_performance import InvalidSourceGrouping
from livelead.domain.reporting.time_window import InvalidDashboardTimeWindow
from livelead.interfaces.auth.tenant_context import TenantContext, get_tenant_context
from livelead.interfaces.rest.deps import get_db_session

router = APIRouter(tags=["reporting"])


@router.get("/reports/export")
async def export_report(
    report_type: str = Query(
        ..., description="dashboard | funnel | source_performance | content_effectiveness"
    ),
    format: str = Query(..., alias="format", description="csv | printable (pdf/html alias)"),
    start: date | None = Query(default=None),
    end: date | None = Query(default=None),
    preset: str | None = Query(default=None),
    grouping: str | None = Query(default=None),
    tenant: TenantContext = Depends(get_tenant_context),
    session: AsyncSession = Depends(get_db_session),
):
    svc = ReportExportService(session)
    try:
        artifact = await svc.export_report(
            tenant.organization_id,
            report_type_raw=report_type,
            format_raw=format,
            start=start,
            end=end,
            preset=preset,
            grouping=grouping,
        )
    except UnsupportedReportExport as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except InvalidDashboardTimeWindow as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except InvalidSourceGrouping as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except InvalidContentGrouping as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return Response(
        content=artifact.body,
        media_type=artifact.media_type,
        headers={
            "Content-Disposition": f'attachment; filename="{artifact.filename}"',
            "X-Report-Type": artifact.report_type.value,
            "X-Export-Format": artifact.export_format.value,
            "X-Generated-At": artifact.generated_at.isoformat(),
        },
    )
